File: Swift/백준/Silver/2167. 2차원 배열의 합/2차원 배열의 합.swift
let nm = readLine()!.split(separator: " ").map { Int($0)! }
let (n, m) = (nm[0], nm[1])

var arr = [[Int]]()
for _ in 0..<n {
    arr.append(readLine()!.split(separator: " ").map { Int($0)! })
}

// 1-based index 누적합 배열 만들기 (n+1) x (m+1)
var prefix = Array(repeating: Array(repeating: 0, count: m + 1), count: n + 1)

for i in 1...n {
    for j in 1...m {
        prefix[i][j] = arr[i-1][j-1]
                    + prefix[i-1][j]
                    + prefix[i][j-1]
                    - prefix[i-1][j-1]
    }
}

let k = Int(readLine()!)!
for _ in 0..<k {
    let input = readLine()!.split(separator: " ").map { Int($0)! }
    let (i, j, x, y) = (input[0], input[1], input[2], input[3])
    
    let sum = prefix[x][y]
            - prefix[i-1][y]
            - prefix[x][j-1]
            + prefix[i-1][j-1]
    
    print(sum)
}

File: Python/백준/Silver/2167. 2차원 배열의 합/2차원 배열의 합.py
import sys
input = sys.stdin.readline

# 배열 크기 입력
N, M = map(int, input().split())

# 원본 배열 입력 (1-based로 다루기 위해 padding)
arr = [[0] * (M+1)]
for _ in range(N):
    row = [0] + list(map(int, input().split()))
    arr.append(row)

# prefix sum 배열
ps = [[0] * (M+1) for _ in range(N+1)]
for i in range(1, N+1):
    for j in range(1, M+1):
        ps[i][j] = arr[i][j] + ps[i-1][j] + ps[i][j-1] - ps[i-1][j-1]

# 쿼리 처리
K = int(input())
out = []
for _ in range(K):
    i, j, x, y = map(int, input().split())
    s = ps[x][y] - ps[i-1][y] - ps[x][j-1] + ps[i-1][j-1]
    out.append(str(s))

print("\n".join(out))